File: src/ovi_voice_assistant/memory/retain_test.py
"""Tests for the retain module."""

from unittest.mock import AsyncMock, MagicMock

import pytest

from ovi_voice_assistant.memory.retain import _build_fact_text, retain
from ovi_voice_assistant.memory.store import MemoryStore


@pytest.fixture
def store():
    s = MemoryStore(":memory:")
    s.open()
    yield s
    s.close()


@pytest.fixture
def embedder():
    mock = AsyncMock()
    mock.embed = AsyncMock(return_value=[[0.1, 0.2, 0.3]])
    return mock


@pytest.fixture
def llm():
    mock = AsyncMock()
    choice = MagicMock()
    choice.message.content = """[
        {
            "what": "Alice is allergic to shellfish",
            "who": "Alice",
            "where": "",
            "when": "",
            "why": "dietary restriction",
            "fact_type": "experience",
            "confidence": 0.95,
            "entities": [{"name": "Alice", "type": "person"}]
        }
    ]"""
    response = MagicMock()
    response.choices = [choice]
    mock.chat.completions.create = AsyncMock(return_value=response)
    return mock


def test_build_fact_text():
    raw = {"what": "Alice likes pizza", "who": "Alice", "where": "NYC", "when": "", "why": ""}
    text = _build_fact_text(raw)
    assert "Alice likes pizza" in text
    assert "who: Alice" in text
    assert "where: NYC" in text


def test_build_fact_text_minimal():
    text = _build_fact_text({"what": "Simple fact"})
    assert text == "Simple fact"


@pytest.mark.asyncio
async def test_retain_extracts_facts(store, embedder, llm):
    result = await retain(
        bank_id="test",
        content="Alice told me she's allergic to shellfish",
        llm=llm,
        llm_model="test-model",
        embedder=embedder,
        store=store,
    )
    assert result.success
    assert result.facts_count == 1
    assert len(result.fact_ids) == 1

    # Verify fact was stored
    facts = store.get_facts("test")
    assert len(facts) == 1
    assert "allergic to shellfish" in facts[0].what


@pytest.mark.asyncio
async def test_retain_extracts_entities(store, embedder, llm):
    result = await retain(
        bank_id="test",
        content="Alice told me she's allergic to shellfish",
        llm=llm,
        llm_model="test-model",
        embedder=embedder,
        store=store,
    )
    assert len(result.entity_ids) == 1

    entities = store.get_entities("test")
    assert len(entities) == 1
    assert entities[0].text == "Alice"


@pytest.mark.asyncio
async def test_retain_empty_extraction(store, embedder, llm):
    choice = MagicMock()
    choice.message.content = "[]"
    response = MagicMock()
    response.choices = [choice]
    llm.chat.completions.create = AsyncMock(return_value=response)

    result = await retain(
        bank_id="test",
        content="Hello, how are you?",
        llm=llm,
        llm_model="test-model",
        embedder=embedder,
        store=store,
    )
    assert result.success
    assert result.facts_count == 0


@pytest.mark.asyncio
async def test_retain_handles_llm_failure(store, embedder, llm):
    llm.chat.completions.create = AsyncMock(side_effect=Exception("API error"))

    result = await retain(
        bank_id="test",
        content="some content",
        llm=llm,
        llm_model="test-model",
        embedder=embedder,
        store=store,
    )
    assert not result.success


@pytest.mark.asyncio
async def test_retain_handles_invalid_json(store, embedder, llm):
    choice = MagicMock()
    choice.message.content = "not valid json"
    response = MagicMock()
    response.choices = [choice]
    llm.chat.completions.create = AsyncMock(return_value=response)

    result = await retain(
        bank_id="test",
        content="some content",
        llm=llm,
        llm_model="test-model",
        embedder=embedder,
        store=store,
    )
    assert not result.success


@pytest.mark.asyncio
async def test_retain_entity_dedup(store, embedder, llm):
    """Retaining twice with same entity should merge fact_ids."""
    # First retain — creates "Alice" entity
    embedder.embed = AsyncMock(return_value=[[0.1, 0.2, 0.3]])
    await retain(
        bank_id="test", content="first", llm=llm, llm_model="m",
        embedder=embedder, store=store,
    )

    # Second retain — "Alice" entity should merge
    embedder.embed = AsyncMock(return_value=[[0.4, 0.5, 0.6]])
    await retain(
        bank_id="test", content="second", llm=llm, llm_model="m",
        embedder=embedder, store=store,
    )

    entities = store.get_entities("test")
    assert len(entities) == 1  # Only one "Alice"
    assert len(entities[0].fact_ids) == 2  # Both facts linked

File: internal/memory/retain_test.go
package memory

import (
	"context"
	"errors"
	"testing"

	"github.com/bryfur/ovi-voice-assistant/internal/llm"
)

type fakeChat struct {
	content string
	err     error
	calls   int
}

func (f *fakeChat) Chat(_ context.Context, _ llm.ChatRequest) (*llm.ChatResponse, error) {
	f.calls++
	if f.err != nil {
		return nil, f.err
	}
	return &llm.ChatResponse{Choices: []llm.Choice{{Message: llm.Message{Content: f.content}}}}, nil
}

const extractJSON = `[
  {
    "what": "Alice is allergic to shellfish",
    "who": "Alice",
    "where": "",
    "when": "",
    "why": "dietary restriction",
    "fact_type": "experience",
    "confidence": 0.95,
    "entities": [{"name": "Alice", "type": "person"}]
  }
]`

func TestBuildFactText(t *testing.T) {
	got := BuildFactText("Alice likes pizza", "Alice", "NYC", "", "")

	if got != "Alice likes pizza. who: Alice. where: NYC" {
		t.Fatalf("got %q", got)
	}
	if BuildFactText("Simple fact", "", "", "", "") != "Simple fact" {
		t.Fatal("minimal text wrong")
	}
}

func TestParseFactsStripsFencesAndAcceptsObject(t *testing.T) {
	facts, err := ParseFacts("```json\n" + extractJSON + "\n```")
	if err != nil || len(facts) != 1 || facts[0].Who != "Alice" {
		t.Fatalf("fenced: %+v, %v", facts, err)
	}

	single, err := ParseFacts(`{"what":"x"}`)

	if err != nil || len(single) != 1 {
		t.Fatalf("single: %+v, %v", single, err)
	}
}

func TestRetainExtractsFactsAndEntities(t *testing.T) {
	s := openStore(t)
	emb := &fakeEmbedder{vec: []float32{0.1, 0.2, 0.3}}

	res, err := Retain(context.Background(), "test", "Alice told me she's allergic to shellfish", &fakeChat{content: extractJSON}, "m", emb, s, "")

	if err != nil || !res.Success || res.FactsCount != 1 || len(res.FactIDs) != 1 || len(res.EntityIDs) != 1 {
		t.Fatalf("got %+v, %v", res, err)
	}
	facts, _ := s.GetFacts("test", nil)
	if len(facts) != 1 || facts[0].FactType != FactExperience || facts[0].Confidence != 0.95 || len(facts[0].Embedding) != 3 {
		t.Fatalf("stored fact = %+v", facts[0])
	}
	entities, _ := s.GetEntities("test")
	if len(entities) != 1 || entities[0].Text != "Alice" || entities[0].EntityType != EntityPerson {
		t.Fatalf("entities = %+v", entities)
	}
}

func TestRetainEmptyExtraction(t *testing.T) {
	s := openStore(t)

	res, err := Retain(context.Background(), "test", "Hello", &fakeChat{content: "[]"}, "m", &fakeEmbedder{}, s, "")

	if err != nil || !res.Success || res.FactsCount != 0 {
		t.Fatalf("got %+v, %v", res, err)
	}
}

func TestRetainHandlesLLMFailureAndInvalidJSON(t *testing.T) {
	s := openStore(t)

	failed, err1 := Retain(context.Background(), "test", "x", &fakeChat{err: errors.New("api")}, "m", &fakeEmbedder{}, s, "")
	invalid, err2 := Retain(context.Background(), "test", "x", &fakeChat{content: "not json"}, "m", &fakeEmbedder{}, s, "")

	if err1 != nil || err2 != nil || failed.Success || invalid.Success {
		t.Fatalf("got %+v/%v %+v/%v", failed, err1, invalid, err2)
	}
}

func TestRetainEntityDedup(t *testing.T) {
	s := openStore(t)
	emb := &fakeEmbedder{vec: []float32{0.1, 0.2, 0.3}}

	Retain(context.Background(), "test", "first", &fakeChat{content: extractJSON}, "m", emb, s, "")
	Retain(context.Background(), "test", "second", &fakeChat{content: extractJSON}, "m", emb, s, "")

	entities, _ := s.GetEntities("test")
	if len(entities) != 1 || len(entities[0].FactIDs) != 2 {
		t.Fatalf("entities = %+v", entities)
	}
}
